### PR TITLE
modified models with jpa

### PR DESCRIPTION
**Pull request by Kasgor**
added changes to all models

### CODE REVIEW

File: src/main/java/com/example/cult_of_tim/cultoftim/models/Book.java [hunk 1]
```diff
@@ -1,16 +1,23 @@
...
 
+import java.util.List;
 
+@Entity
 public class Book {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
     private Long id;
 
     private String title;
```
Comment (Anoromi) [inline src/main/java/com/example/cult_of_tim/cultoftim/models/Book.java:7]: Варто додати макимальний розмір.

File: src/main/java/com/example/cult_of_tim/cultoftim/models/Category.java [hunk 2]
```diff
@@ -1,10 +1,28 @@
...
+import java.util.List;
+
+@Entity
 public class Category {
 
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
     private Long id;
+
     private String name;
```
Comment (Anoromi) [inline src/main/java/com/example/cult_of_tim/cultoftim/models/Category.java]: Розмір

File: src/main/java/com/example/cult_of_tim/cultoftim/models/Promotion.java [hunk 3]
```diff
@@ -1,27 +1,37 @@
...
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
     private Long id;
+
     private String description;
     private LocalDateTime startDate;
     private LocalDateTime endDate;
-    private List<Long> userIDs;
+
+    @ManyToMany
+    private List<User> users;
```
Comment (Anoromi) [inline src/main/java/com/example/cult_of_tim/cultoftim/models/Promotion.java]: До чого у Promotion з'єднання до користувачів?

File: cult-of-tim-auth/src/main/java/com/cult_of_tim/auth/cultoftimauth/repositories/UserRepository.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+public interface UserRepository extends JpaRepository<User, UUID> {
+
+    Optional<User> findById(UUID id);
+    Optional<User> findByUsername(String username);
+
+    Optional<User> findByEmail(String email);
+
+    User save(User user);
+
+    void deleteById(UUID id);
```
Comment (HoWurk) [inline cult-of-tim-auth/src/main/java/com/cult_of_tim/auth/cultoftimauth/repositories/UserRepository.java]: Most of these methods exist in jparepository and don't need to be duplicated:
image
image
etc
Other Repository files, same thing

File: src/main/java/com/example/cult_of_tim/cultoftim/service/impl/PromotionMockService.java [hunk 5]
```diff
@@ -12,106 +13,86 @@
...
-        return promotionDao.isGlobal(promotionID);
+        return promotionRepository.existsByIdAndGlobal(promotionID, true);
     }
 
     @Override
-    public Long createPromotion(Promotion promotion) {
-        return promotionDao.createPromotion(promotion);
+    public Promotion createPromotion(Promotion promotion) {
+        return promotionRepository.save(promotion);
     }
```
Comment (HoWurk) [inline src/main/java/com/example/cult_of_tim/cultoftim/service/impl/PromotionMockService.java]: We receive not an entity here, need to change to PromotionCreationDTO or similar stuff
https://www.baeldung.com/java-dto-pattern

File: src/main/java/com/example/cult_of_tim/cultoftim/service/impl/PromotionMockService.java [hunk 6]
```diff
@@ -12,106 +13,86 @@
...
 
     @Override
     public Promotion updatePromotion(Long id, Promotion updatedPromotion) {
         updatedPromotion.setId(id);
-        return promotionDao.updatePromotion(updatedPromotion);
+        return promotionRepository.save(updatedPromotion);
```
Comment (HoWurk) [inline src/main/java/com/example/cult_of_tim/cultoftim/service/impl/PromotionMockService.java]: same as create, need possibly update dto

File: src/main/java/com/example/cult_of_tim/cultoftim/models/Author.java [hunk 7]
```diff
@@ -1,9 +1,24 @@
...
-    private String name;
+
+    @Column(length = 100)
+    private String firstName;
+
+    @Column(length = 100)
+    private String lastName;
+
+    @ManyToMany
+    private List<Book> books;
```
Comment (HoWurk) [inline src/main/java/com/example/cult_of_tim/cultoftim/models/Author.java]: Either make author contain books, or books contain authors. One of the two.
Same with categories and books.

File: cult-of-tim-auth/src/main/java/com/cult_of_tim/auth/cultoftimauth/dao/UserDao.java [hunk 8]
```diff
@@ -11,17 +11,17 @@
...
 
-    Long createUser(User user);
+    UUID createUser(User user);
 
     User updateUser(User user);
 
-    void deleteUserById(Long id);
+    void deleteUserById(UUID id);
 
     Optional<UserToken> getUserTokenFor(UUID token);
```
Comment (HoWurk) [inline cult-of-tim-auth/src/main/java/com/cult_of_tim/auth/cultoftimauth/dao/UserDao.java]: If we move on from UserDao, we either change it fully with UserDaoMock and UserMockService so it actually builds, or don't change it, but still fix it.

File: src/main/java/com/example/cult_of_tim/cultoftim/service/impl/PromotionMockService.java [hunk 9]
```diff
@@ -37,17 +38,29 @@ public List<Promotion> getUserPromotions(Long userID) {
 
     @Override
     public boolean isGlobal(Long promotionID) {
-        return promotionRepository.existsByIdAndGlobal(promotionID, true);
+        return promotionRepository.existsByIdAndGlobalPromotion(promotionID, true);
```
Comment (HoWurk) [inline src/main/java/com/example/cult_of_tim/cultoftim/service/impl/PromotionMockService.java]: First, change globalPromotion to global (boolean in Promotion).
Second, make method better in PromotionRepository, read comment there.
Third, insert here. If for some reason you need existsbysomething, don't put it in isGlobalMethod.

File: src/main/java/com/example/cult_of_tim/cultoftim/service/PromotionService.java [hunk 10]
```diff
@@ -15,9 +16,9 @@ public interface PromotionService {
 
     boolean isGlobal(Long promotionID);
 
-    Promotion createPromotion(Promotion promotion);
+    Promotion createPromotion(String description, LocalDateTime startDate, LocalDateTime endDate, List<PromotionDiscount> discounts, boolean globalPromotion);
 
-    Promotion updatePromotion(Long id, Promotion updatedPromotion);
+    Promotion updatePromotion(Long id, String description, LocalDateTime startDate, LocalDateTime endDate, List<PromotionDiscount> discounts, boolean globalPromotion);
```
Comment (HoWurk) [inline src/main/java/com/example/cult_of_tim/cultoftim/service/PromotionService.java]: this is just plain wrong, again, https://www.baeldung.com/java-dto-pattern
Or leave it as it was previously, and ask lecturer next time.

File: src/main/java/com/example/cult_of_tim/cultoftim/repositories/PurchaseRepository.java [hunk 11]
```diff
@@ -10,13 +10,13 @@
...
-    List<Purchase> findAll();
+    //List<Purchase> findAll();
 
     List<Purchase> findByUserId(Long userId);
 
-    Purchase save(Purchase purchase);
+    //Purchase save(Purchase purchase);
 
-    void deleteById(Long id);
+    //void deleteById(Long id);
```
Comment (HoWurk) [inline src/main/java/com/example/cult_of_tim/cultoftim/repositories/PurchaseRepository.java]: why comment and not just delete?

File: src/main/java/com/example/cult_of_tim/cultoftim/repositories/PromotionRepository.java [hunk 12]
```diff
@@ -10,15 +10,15 @@
...
-    Optional<Promotion> findById(Long id);
+    //Optional<Promotion> findById(Long id);
 
-    List<Promotion> findAll();
+    //List<Promotion> findAll();
 
     List<Promotion> findByUserId(Long userId);
 
-    boolean existsByIdAndGlobal(Long promotionId, boolean global);
+    boolean existsByIdAndGlobalPromotion(Long promotionId, boolean global);
```
Comment (HoWurk) [inline src/main/java/com/example/cult_of_tim/cultoftim/repositories/PromotionRepository.java]: I don't know why this is needed, if you need to check if item exists, then just do boolean exists, if you need to check if this id has global true,
boolean existsByIdAndGlobalPromotionTrue(Long promotionId);
https://www.baeldung.com/spring-data-derived-queries

File: src/main/java/com/example/cult_of_tim/cultoftim/service/impl/PromotionMockService.java [hunk 13]
```diff
@@ -37,17 +38,29 @@ public List<Promotion> getUserPromotions(Long userID) {
...
-    public Promotion updatePromotion(Long id, Promotion updatedPromotion) {
+    public Promotion updatePromotion(Long id, String description, LocalDateTime startDate, LocalDateTime endDate, List<PromotionDiscount> discounts, boolean globalPromotion) {
+        Promotion updatedPromotion = new Promotion();
         updatedPromotion.setId(id);
+        updatedPromotion.setDescription(description);
+        updatedPromotion.setStartDate(startDate);
+        updatedPromotion.setEndDate(endDate);
+        updatedPromotion.setDiscounts(discounts);
+        updatedPromotion.setGlobalPromotion(globalPromotion);
         return promotionRepository.save(updatedPromotion);
```
Comment (HoWurk) [inline src/main/java/com/example/cult_of_tim/cultoftim/service/impl/PromotionMockService.java]: change it back, or fix it, read in interface comment

File: src/main/java/com/example/cult_of_tim/cultoftim/repositories/PromotionRepository.java [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+import java.util.Optional;
+
+@Repository
+public interface PromotionRepository extends JpaRepository<Promotion, Long> {
+
+    //Optional<Promotion> findById(Long id);
+
+    //List<Promotion> findAll();
+
+    List<Promotion> findByUserId(Long userId);
```
Comment (HoWurk) [inline src/main/java/com/example/cult_of_tim/cultoftim/repositories/PromotionRepository.java]: still doesn't work, no field userId in promotion. if you want to query, @Query

File: src/main/java/com/example/cult_of_tim/cultoftim/repositories/BookRepository.java [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+
+import java.util.List;
+import java.util.Optional;
+
+@Repository
+public interface BookRepository extends JpaRepository<Book, Long> {
+
+    List<Book> findByAuthorId(Long authorId);
+
+    List<Book> findByCategoryId(Long categoryId);
```
Comment (HoWurk) [inline src/main/java/com/example/cult_of_tim/cultoftim/repositories/BookRepository.java]: still doesnt work, findAllByAuthorsContaining, same with category
no fields authorId, categoryId

**Review by HoWurk — CHANGES_REQUESTED** [review]

**Review by HoWurk — APPROVED** [review]

**Pull request merged**